Polish peek view for inline reviews

donokuda opened the pull request:
This pull request starts bringing in support for themes to inline reviews.

Doesn't look 100% great so I'll be following up with some custom color choices in a series of future pull requests. Here's what it looks like so far:

**Light**
<img width="400" src="https://user-images.githubusercontent.com/1174461/28340539-fec353ee-6bc4-11e7-8600-36dfb1cb48fd.PNG" />

**Dark**
<img src="https://user-images.githubusercontent.com/1174461/28340537-febfe4ca-6bc4-11e7-8474-48352b183037.PNG" width="400" />

**Blue**
<img src="https://user-images.githubusercontent.com/1174461/28340631-4b2e4e0a-6bc5-11e7-868c-e777b5d35492.PNG" width="400" />

jcansdale commented on src/GitHub.InlineReviews/Tags/ShowInlineCommentGlyph.xaml:
```diff
@@ -4,21 +4,38 @@
+        <Style TargetType="UserControl">
+            <Style.Triggers>
+                <DataTrigger Binding="{Binding DiffChangeType}" Value="Add">
+                    <Setter Property="Background" Value="{DynamicResource DiffChangeBackground.Add}" />
```
I've created 3 new dynamic brushes that will change according the the theme. They're called `DiffChangeBackground.Add`, `DiffChangeBackground.Delete` and `DiffChangeBackground.None` to mirror the `DiffChangeType` enum.

The old `DiffChangeBackground` is no longer being set for each glyph. Everything is now done via the `DiffChangeType` property.

jcansdale commented:
I noticed the arrows won't showing up here:
![image](https://user-images.githubusercontent.com/11719160/28424012-7b229c56-6d64-11e7-8592-1b80771593f5.png)

Strange, because they show in your screen grabs. I did merge from `master`. Maybe something got messed up in the process?

jcansdale commented on src/GitHub.InlineReviews/Tags/ShowInlineCommentGlyph.xaml:
```diff
@@ -4,21 +4,38 @@
+
+    <UserControl.ToolTip>
+        <ToolTip
+            Background="{DynamicResource GitHubVsToolWindowBackground}"
```
You can now access the GitHub dynamic/themed resources from the glyph XAML files (this wouldn't have worked before). You can see I'm using `GitHubVsToolWindowBackground` here.

jcansdale commented on src/GitHub.InlineReviews/Properties/DesignTimeResources.xaml:
```diff
@@ -0,0 +1,15 @@
+        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+
+    <SolidColorBrush x:Key="DiffChangeBackground.Add" Color="Green" />
```
If you want any resources to show up in the designer, you can add there here.

You can see below, I'm also adding the 3 `SharedDictionary.xaml` files. At the moment this is only needed for the `Show/AddInlineCommentGlyph.xaml` files (because they're added by `BrushesManager.cs` rather then an explicit `MergedDictionaries` element in each XAML file).

donokuda commented:
>Strange, because they show in your screen grabs. I did merge from master. Maybe something got messed up in the process?

🤔 Hmm, that is strange.. I'll take a lookie loo at this.

donokuda commented:
Hmm, not running into that problem here...

![image](https://user-images.githubusercontent.com/1174461/28471913-90a23e6c-6df3-11e7-84bf-e008eccba8fd.png)

donokuda commented:
Here's where I'm at right now:

![blue](https://user-images.githubusercontent.com/1174461/28485057-8b24413c-6e2b-11e7-8082-a224ed98c090.png)
![dark](https://user-images.githubusercontent.com/1174461/28485059-8b29a88e-6e2b-11e7-9af5-732a2693d822.png)
![light pn](https://user-images.githubusercontent.com/1174461/28485058-8b24e0e2-6e2b-11e7-89af-e2e2193a7e90.PNG)

Most of the focus is around the margin in these screenshots (nothing has changed for the comment boxes.)

This is ready for review 🙏

jcansdale commented:
@donokuda,

At the moment there are a whole bunch of different shades of red/green on the diff view. I wonder if the comment bubbles could borrow the colors from the scroll bar indicators?

![image](https://user-images.githubusercontent.com/11719160/28516875-91c17e16-705a-11e7-997b-cf558fa39ac1.png)

I think the colors might work and make it look a little less busy? What do you think?

jcansdale commented:
When I was taking a bunch of screen-grabs for the tooltip issue, I noticed that I mistook the tracking comment for a real comment on a number of occasions. I wondered if there might be a might be a way to de-emphasize tracking comment glyph (with the `+`) to make this less likely to happen?

![image](https://user-images.githubusercontent.com/11719160/28517095-39bccfee-705b-11e7-842e-347cb8bed748.png)

donokuda commented:
> At the moment there are a whole bunch of different shades of red/green on the diff view. I wonder if the comment bubbles could borrow the colors from the scroll bar indicators?

I like this idea and will give it a shot.

> When I was taking a bunch of screen-grabs for the tooltip issue, I noticed that I mistook the tracking comment for a real comment on a number of occasions. I wondered if there might be a might be a way to de-emphasize tracking comment glyph (with the +) to make this less likely to happen?

Yup. I think one idea is to use a square with the plus instead like what we do on dotcom:

<img src="https://user-images.githubusercontent.com/1174461/28535998-2bd82b80-705b-11e7-9e25-55492842a608.png" width="100" />

donokuda commented:
New fill colors added:

**Light**
![image](https://user-images.githubusercontent.com/1174461/28542511-40472696-7071-11e7-8f3a-404e8c9a0179.png)

**Blue**
![image](https://user-images.githubusercontent.com/1174461/28542582-8350206e-7071-11e7-834b-ad66c25df018.png)

**Dark**
![image](https://user-images.githubusercontent.com/1174461/28542622-a8ddcba6-7071-11e7-8ab6-bfec8be95844.png)

donokuda commented:
Changed the "Add comment" glyph to a square icon and adjusted the blue for the dark theme.

**Dark**
![image](https://user-images.githubusercontent.com/1174461/28547904-02e72c5a-7086-11e7-885a-4e7494a6d441.png)

**Blue**
![image](https://user-images.githubusercontent.com/1174461/28547929-1ea8405a-7086-11e7-91c1-b887470db806.png)

**Light**
![image](https://user-images.githubusercontent.com/1174461/28547951-32ec13a2-7086-11e7-8292-84a939fe30b5.png)

jcansdale commented:
> I think one idea is to use a square with the plus instead like what we do on dotcom

I've been having a play with this iteration. I have a few reservations.

- Having the icon change color as it tracks up and down can look a bit psychedelic
- Because the `+` isn't visible when there's already a comment, it flashes on and off
- The add comment glyph is now more prominent than the comment here glyph

Do you think showing just the `+` might work?

jcansdale commented:
> One thing to point out: the different bg colored margin indicates that the line can be commented on. It does not imply that the line has changed in that branch. So even with the goal of "When I run into a bug while working in a branch, I want to know which lines have changed so I understand if I introduced the bug" the implementation as it stands doesn't hold up to that goal.

I realized we hadn't talked about this and it is an important part of the equation. Here are some pertinent notes from the mini-summit (I'll discuss in a separate comment):

# Margin

We're currently displaying the comment icons in the glyph margin, but this has a number of problems:

- We interfere with existing glyph margin functionality such as breakpoints
- We need to create a glyph for every line where a comment can be left
- It's not clear on which lines comments can be left

The plan is to add our own margin which will consist of:

- Icons for existing comments
- A method of visually indicating where comments can be left
- When the mouse is hovered over a line where a comment can be left, an "add comment" icon will appear

Regarding point 2 - in future when we move to the new GraphQL reviews API then it will be possible to leave comments anywhere, so rather than say highlighting the lines where comments can be left we felt it better to highlight lines where comments _can't_ be left - this way once the restriction is removed there will be no re-learning the interface for the user. We thought that maybe a subtle crosshatch or strikeout pattern in the margin might be the best visual indication for "you can't leave comments here".

A rough mockup of what I'm envisioning, where the cross-hatched areas represent areas that can't be commented on:

![image](https://cloud.githubusercontent.com/assets/1775141/26449256/7dbee71a-4151-11e7-9a51-481ad2390f1b.png)

jcansdale commented:
> the different bg colored margin indicates that the line can be commented on. It does not imply that the line has changed in that branch.

At the moment only lines within 3 lines of a change can be commented on (in its current incarnation, but this is likely to change at some point). In the mockup above, the cross-hatched areas represent areas that can't be commented on.

Without any other visual cues, the non-cross-hatched areas do absolutely imply that a line has changed in that branch. The user can see that one or more lines have changed, but it isn't clear which lines have changed. Hence the need to disambiguate changed lines from commentable lines.

I'm be happy with a subtle que indicating which lines have changed (doesn't have to be red/green). This will become all the more important if/when all lines become commentable.

shana commented:
I think we should probably focus on the UI for the diff view only for now, and consider the UI for the editor itself to be a separate case - we've already seen that they have different context and different use cases, and considering them together is increasingly complicating things. We should have a separate issue for discussing the design of the inline comments UI in the editor itself, I think.

So, theming! In .com we always show a blue comment icon regardless of the type of change that happened in the line. What do you think of doing the same here @donokuda?

jcansdale commented:
> So, theming! In .com we always show a blue comment icon regardless of the type of change that happened in the line. What do you think of doing the same here @donokuda?

This is basically what I prototyped in the latest commit (at least in the light themes). I stopped the glyphs from differentiating between their added/deleted line status. @donokuda can back this out if he wants but I wanted to dogfood it for a while. Having used it for a bit I do prefer the more sober colors.

I think we still need to consider how best to represent non-commentable lines (keeping in mind they'll probably go away eventually). In this latest commit, I found it could be confusing when there's equal mix of commentable and non-commentable lines.  Anyway, just an observation.

shana commented on src/GitHub.InlineReviews/Tags/AddInlineCommentGlyph.xaml:
```diff
@@ -2,36 +2,55 @@
+        </Style>
+    </UserControl.Style>
+
+    <Grid>
```
Just doing a little comment to see how it looks...

shana replied:
... and replying to myself in the extension!

shana commented:
@donokuda master merged in. this is how it looks to me right now:

![image](https://user-images.githubusercontent.com/310137/28983345-a4e7aea4-7959-11e7-97c7-81a32f5437eb.png)

donokuda commented:
@shana Thank you! 🙇 

Looks like there's some styling stuff I need to fix but it should be pretty straightforward

<img src="https://user-images.githubusercontent.com/1174461/28984457-e799b37e-7912-11e7-9d20-0d19872a586f.png" width="400" />

Or at least more straightforward than me attempting to fix merge conflicts 😄

grokys commented:
Just taking a look at this now: looks great in general, a few things though:

- The white background to the peek view doesn't look great on the blue theme. Could we use the gray that the light theme uses?
- On the light theme, the text box background is the same gray as the background and so the text box looks disabled
- I think we still need some visual indication of where you can leave comments. There's a gray background on lines that were add/removed but IMO it would be better if that were removed and used for lines that can't be commented on

grokys commented:
@donokuda merged master which now includes #1112 for allowing for the peek view to auto-size to its content.

donokuda commented:
> - The white background to the peek view doesn't look great on the blue theme. Could we use the gray that the light theme uses?
> - On the light theme, the text box background is the same gray as the background and so the text box looks disabled

Good calls on both. For the textbox, I updated to use the searchbox background color that VS provides us. Here's what the peek view looks like right now:

<img src="https://user-images.githubusercontent.com/1174461/29037151-47764316-7b57-11e7-9a82-58c9b6e239f5.png" width="500" />

<img src="https://user-images.githubusercontent.com/1174461/29037198-7527b420-7b57-11e7-8a6d-5d99242eb179.png" width="500" />

<img src="https://user-images.githubusercontent.com/1174461/29037230-935aaf38-7b57-11e7-9fa5-77f56ffd542f.png" width="500" />

donokuda commented:
> I think we still need some visual indication of where you can leave comments. There's a gray background on lines that were add/removed but IMO it would be better if that were removed and used for lines that can't be commented on

Updated the margins:

<img src="https://user-images.githubusercontent.com/1174461/29039846-c2e35c1e-7b61-11e7-8224-3330b34e483f.png" width="450" />


<img src="https://user-images.githubusercontent.com/1174461/29039899-f98e4594-7b61-11e7-9813-e9605c068543.png" width="450" />


<img src="https://user-images.githubusercontent.com/1174461/29039876-e5b35104-7b61-11e7-98f7-b04ab584e873.png" width="450" />

donokuda commented:
Updated codeblock styles:

![image](https://user-images.githubusercontent.com/1174461/29049160-36447b6c-7b89-11e7-8f62-d15dba9070b4.png)

![image](https://user-images.githubusercontent.com/1174461/29049185-624ea25a-7b89-11e7-9e69-f535dc693676.png)

![image](https://user-images.githubusercontent.com/1174461/29049207-7d479dfa-7b89-11e7-936c-ad5f5c777418.png)

@grokys @jcansdale I think this is ready for another review. This branch now has better support for themes and fixes up the peek view so that the experience is better than it was before (by removing and reduce the sizes of some elements.)

grokys commented on src/GitHub.InlineReviews/Tags/AddInlineCommentGlyph.xaml:
```diff
@@ -2,16 +2,55 @@
+        <Border Background="{DynamicResource GitHubGlyphMarginCommentableBackground}" BorderThickness="0,0,1,0" />
+        <Viewbox x:Name="AddViewbox" Margin="1,1,0,0">
+            <Canvas Width="14" Height="14">
+                <Rectangle Width="13" Height="13" />
```
Is this `Rectangle` needed?

donokuda replied:
Yup! The background is being set with a style.

grokys commented on src/GitHub.InlineReviews/Tags/AddInlineCommentGlyph.xaml:
```diff
@@ -2,16 +2,55 @@
+
+                <Canvas.Resources>
+                    <Style TargetType="Rectangle">
+                        <Style.Triggers>
```
Are these triggers still used? AFAICS the add glyph is now always the same color?

grokys commented on src/GitHub.InlineReviews/Tags/ShowInlineCommentGlyph.xaml:
```diff
@@ -3,18 +3,54 @@
+                <Canvas.Resources>
+                    <Style TargetType="Path">
+                        <Style.Triggers>
+                            <DataTrigger Binding="{Binding DiffChangeType}" Value="Add">
```
Again, are these triggers needed?

grokys commented on src/GitHub.InlineReviews/Views/GlyphMarginGrid.xaml:
```diff
@@ -4,67 +4,17 @@
-        -->
-        
-    </Grid.Background>
+    <Border Background="{DynamicResource GitHubPeekViewBackground}" BorderBrush="{DynamicResource GitHubPeekViewBackground}" BorderThickness="0,0,1,0"/>
```
Do we need to specify a border when the background is the same color?

donokuda replied:
Hmm, maybe not. I'll check on this.

donokuda commented:
Here's what I got so far as of 2ba07ba. Good call on giving a little more spacing between the author information and the comment.

![image](https://user-images.githubusercontent.com/1174461/29084260-46738e8c-7c20-11e7-8ce5-3e523dc09e69.png)

donokuda commented:
@grokys Ready for another 👀
